code/game/objects/structures/safe.dm

```diff
@@ -224,20 +309,60 @@ FLOOR SAFES
 /obj/structure/safe/examine_status(mob/user)
 	return
 
+/obj/structure/safe/Destroy()
+	GLOB.safes -= src
+	drill.soundloop.stop()
+	return ..()
+
+/obj/structure/safe/process()
+	if(drill_timer)
+		overlays -= bar
+		bar = image('icons/effects/progessbar.dmi', src, "prog_bar_[round((((world.time - drill_start_time) / time_to_drill) * 100), 5)]", HUD_LAYER)
+		overlays += bar
+		if(prob(15))
+			drill.spark_system.start()
+
+/obj/structure/safe/examine(mob/user)
+	..()
+	if(open)
+		to_chat(user, "On the inside of the the door is <b>[combo_to_open]</b>")
+
 //FLOOR SAFES
 /obj/structure/safe/floor
 	name = "floor safe"
 	icon_state = "floorsafe"
 	density = 0
 	level = 1	//underfloor
 	layer = 2.5
-
+	drill_x_offset = -1
+	drill_y_offset = 20
 
 /obj/structure/safe/floor/Initialize()
 	..()
 	var/turf/T = loc
 	hide(T.intact)
 
-
 /obj/structure/safe/floor/hide(var/intact)
-	invisibility = intact ? 101 : 0
\ No newline at end of file
+	invisibility = intact ? 101 : 0
+
+/obj/item/safe_internals
+	name = "safe internals"
+	desc = "The mechanism and locking bolts for a Scarborough Arms - 2 tumbler safe"
+	icon_state = "safe_internals"
+
+/obj/item/paper/safe_code
+	name = "safe codes"
+	var/owner
+	info = "<div style='text-align:center;'><img src='ntlogo.png'><center><h3>Safe Codes</h3></center>"
+
+/obj/item/paper/safe_code/New()
+	..()
+	addtimer(CALLBACK(src, .proc/populate_codes), 10)
+
+/obj/item/paper/safe_code/proc/populate_codes()
+	for(var/obj/structure/safe/S in GLOB.safes)

```

Comment:
The only thing in `GLOB.safes` will be objects of `/obj/structure/safe` so you can safely use a typeless loop here.

Reply:
All typechecks add up. It'll be minimal because there will be few safes on the map but what's the point of having a typecheck here when they're all safes?